src/dbnode/persist/fs/files.go

```diff
@@ -1231,7 +1199,62 @@ func SortedCommitLogFiles(commitLogsDir string) ([]string, error) {
 	return sortedCommitLogFiles(commitLogsDir, commitLogFilePattern)
 }
 
+type filesetFile struct {
+	volumeIndex int
+	blockStart  xtime.UnixNano
+	fileName    string
+}
+
 type toSortableFn func(files []string) sort.Interface
+type toBlockStartAndVolumeIndexFn func(file string) (xtime.UnixNano, int, error)
+type sortedFilesetFiles []filesetFile
+
+func (s sortedFilesetFiles) Len() int {
+	return len(s)
+}
+
+func (s sortedFilesetFiles) Less(i, j int) bool {
+	ti := s[i].blockStart
+	tj := s[j].blockStart
+
+	if ti.Before(tj) {
+		return true
+	}
+
+	ij := s[j].volumeIndex
+	ii := s[i].volumeIndex
```

Comment:
Nit: I'd give more readable names to these vars, eg. `iStart/jStart, iVolume/jVolume`.